### ISSUE TITLE
关于TRPO车杆环境运行结果

### ISSUE THREAD

**Issue by 24krab** [boyu-ai/Hands-on-RL#95]
我直接copy了trpo这一章节的代码运行结果，但在车杆环境下和教材里展示的结果差距显著：每轮迭代500采样序列的参数下，几乎无法达到200的回报；即使我尝试增大序列数到1000，效果能够有所改善，但还是和展示结果差距明显。在两台电脑上都跑出了类似的结果，请问这可能是什么原因导致的？
![车杆2轮平滑](https://github.com/user-attachments/assets/278d87f2-88b2-4bc5-b0ae-d913569953fb)
![序列1000车杆平滑](https://github.com/user-attachments/assets/3db1d827-a821-4e6b-9efa-0fdb664f6d43)

**Comment by itera-del** [comment]
+1

**Comment by YanikZ35** [comment]
可以试着增加共轭梯度下降循环部分的轮数，实测中10轮也就到0.0几的水平，根本达不到1e-10阈值的水平，我这里改成100轮之后就可以达到200的奖励了，推测时轮次不够导致的计算结果不准确，下附对比图

![Image](https://github.com/user-attachments/assets/a5fe0846-c0d0-4114-b7f0-72bd170d723d)
![Image](https://github.com/user-attachments/assets/5e0d22cc-0722-4b06-b4d3-73646beba9fd)

**Comment by YanikZ35** [comment]
可以试着增加共轭梯度法循环部分的轮数
    def conjugate_gradient(self, grad, states, old_action_dists):  # 共轭梯度法求解方程
        x = torch.zeros_like(grad)
        r = grad.clone()
        p = grad.clone()
        rdotr = torch.dot(r, r)
        for i in range(100):  # 共轭梯度主循环，10步或许有点少，搜索结果不甚准确，给到100发现有较好的性能提升
            Hp = self.hessian_matrix_vector_product(states, old_action_dists, p)
            alpha = rdotr / torch.dot(p, Hp)
            x += alpha * p
            r -= alpha * Hp
            new_rdotr = torch.dot(r, r)
            if new_rdotr < 1e-10:
                break
            beta = new_rdotr / rdotr
            p = r + beta * p
            rdotr = new_rdotr
        return x
实测中10轮也就到0.0几的水平，根本达不到1e-10阈值的水平，我这里改成100轮之后就可以达到200的奖励了，推测是轮次不够导致的计算结果不准确，下附对比图
![Image](https://github.com/user-attachments/assets/3225d67b-d2f9-486f-a4e7-c6b705129489)
![Image](https://github.com/user-attachments/assets/d3aafc30-30ca-4177-8cba-be02befc5b26)

**Comment by 24krab** [comment]
> 可以试着增加共轭梯度法循环部分的轮数 def conjugate_gradient(self, grad, states, old_action_dists): # 共轭梯度法求解方程 x = torch.zeros_like(grad) r = grad.clone() p = grad.clone() rdotr = torch.dot(r, r) for i in range(100): # 共轭梯度主循环，10步或许有点少，搜索结果不甚准确，给到100发现有较好的性能提升 Hp = self.hessian_matrix_vector_product(states, old_action_dists, p) alpha = rdotr / torch.dot(p, Hp) x += alpha * p r -= alpha * Hp new_rdotr = torch.dot(r, r) if new_rdotr < 1e-10: break beta = new_rdotr / rdotr p = r + beta * p rdotr = new_rdotr return x 实测中10轮也就到0.0几的水平，根本达不到1e-10阈值的水平，我这里改成100轮之后就可以达到200的奖励了，推测是轮次不够导致的计算结果不准确，下附对比图 ![Image](https://github.com/user-attachments/assets/3225d67b-d2f9-486f-a4e7-c6b705129489) ![Image](https://github.com/user-attachments/assets/d3aafc30-30ca-4177-8cba-be02befc5b26)

谢谢你的分享！虽然我试着按照你的建议把共轭梯度循环数量修改至100，但效果仍然不理想，此外训练速度明显变慢。不过我之后又尝试把kl散度的约束调整为了0.001（即kl_constraint = 0.001），同时把共轭梯度循环数修改至50，现在得到了较为理想的结果，如图所示。

![Image](https://github.com/user-attachments/assets/671f11bc-adc6-4702-b78e-833ddf8e5fb3)

![Image](https://github.com/user-attachments/assets/9b8d8fee-a2ae-47cc-938b-00e21ee3e9b3)

**Comment by 24krab** [comment]
> +1

我通过修改一部分超参数得到了较为良好的结果，见这个issue里我上一条回复

**Comment by 11443** [comment]
但为何无法复现出相同的结果呢，即使我把gym库的版本设置为书中所用

**Comment by shihaofu2003** [comment]
针对 CartPole-v1 的环境，即最长步数为 500，最大累积奖励( gamma = 1）为 500的情况，建议调整:
* 模型 $hidden\\_dim=512$ 
* 共轭梯度法（conjugate gradient method）的 $num\\_iters=15$。

结果如下：收敛快速且稳定
<img src="https://raw.githubusercontent.com/shihaofu2003/MyPic/imgs/images/20260105123143.png" width="50%">

> > +1
> 
> 我通过修改一部分超参数得到了较为良好的结果，见这个issue里我上一条回复
